### PR TITLE
fix(core-graphql): sorting and limiting of records

### PR DESCRIPTION
**Pull request by JeremiGendron**
## Proposed changes

1. The Transaction type is the only one where the address field (recipientId) has the Address Scalar type, this is unwanted because every other type uses Strings for address property values. The fix is to remove the erroneous Address Scalar typedef and resolvers, which didn't seem to work either.

2. When using orderBy, if mentioning `ASC` as the value it will not be recognized as an AST node by the resolver (tested on transactions endpoint). The reason why is because the database takes queries for ordering in lowercase format (`desc || asc`) only and the helper doesn't convert to lowercase. Easy fix is to convert to lowercase all orderBy direction values (done).

Also instates a collection of graphql tests, most of which work. For transactions, there is still a bug that needs to be figured out where the limit param doesn't get taken into account. I suspect this is because of some hardcoded 100 limit. There was also an issue with `wallets.test.js`, where the limit parameter also is erroneous:  `database.createCondition is not a function`

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improve a current implementation without adding a new feature or fixing a bug)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build (changes that affect the build system)
- [ ] Docs (documentation only changes)
- [x] Test (adding missing tests or fixing existing tests)
- [ ] Other... Please describe:

## Checklist

- [x] I have read the [CONTRIBUTING](https://docs.ark.io/developers/guidelines/contributing.html) documentation
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by JeremiGendron** [comment]
closes #1268

**Comment by JeremiGendron** [comment]
Also added: The transaction type test was failing hard because it passed the string value 'TRANSFER' to the DB query, which accepts an Int value; easy fix replace `parameters.type` in the repository (query module for transactions) to `TRANSACTION_TYPES[parameters.type]` in-place. Works good :+1:

**Comment by JeremiGendron** [comment]
Another nasty bug: the resolver for transactions didn't pass the limit to the repository query (db), so it was always undefined and defaulted to 100. Fixed.

**Comment by JeremiGendron** [comment]
Issue potentially averted: the default orderBy string for the wallets resolver is 'height:DESC', this is problematic in many cases as shown with the transactions repository example where the orderBy string is not properly formatted for a DB query.

File: packages/core-graphql/__tests__/api/wallets.test.js [hunk 1]
```diff
@@ -66,10 +66,39 @@ describe('GraphQL API { wallets }', () => {
   })
 
   describe.skip('GraphQL queries for Wallets - testing relationships', () => {
+    it('should verify that relationships are valid', async () => {
+      const query = '{ wallets(limit: 1) { transactions { id } } }'
+      const response = await utils.request(query)
+
+      expect(response).toBeSuccessfulResponse()
 
+      console.log(response.data)
```
Comment (faustbrian) [inline packages/core-graphql/__tests__/api/wallets.test.js]: Should be either removed or be replaced by an assertion.

File: packages/core-graphql/__tests__/api/wallets.test.js [hunk 2]
```diff
@@ -66,10 +66,39 @@ describe('GraphQL API { wallets }', () => {
...
+      /**
+       *  { data: { wallets: [ [Object] ] },
+            errors: [ { message: 'database.createCondition is not a function',
+              locations: [Array],
+              path: [Array],
+              extensions: [Object] } ]
+          }
+       */
+
+      console.log(response.data.wallets)
```
Comment (faustbrian) [inline packages/core-graphql/__tests__/api/wallets.test.js]: Should be either removed or be replaced by an assertion.

File: packages/core-graphql/lib/defs/inputs.js [hunk 3]
```diff
@@ -12,7 +12,6 @@ module.exports = `
   scalar JSON
   scalar Limit
   scalar Offset
-  scalar Address
```
Comment (faustbrian) [inline packages/core-graphql/lib/defs/inputs.js]: Address is the correct type. You can find it in `core-graphql/lib/resolvers/index.js`.

File: packages/core-graphql/lib/defs/inputs.js [hunk 4]
```diff
@@ -35,7 +34,7 @@ module.exports = `
     fee: Float
     blockId: String
     senderPublicKey: String
-    recipientId: Address
+    recipientId: String
```
Comment (faustbrian) [inline packages/core-graphql/lib/defs/inputs.js]: Address is the correct type. You can find it in `core-graphql/lib/resolvers/index.js`.

File: packages/core-graphql/lib/repositories/transactions.js [hunk 5]
```diff
@@ -47,6 +51,7 @@ class TransactionsRepository extends Repository {
 
     applyConditions([selectQuery, countQuery])
 
+console.log(parameters.limit)
```
Comment (faustbrian) [inline packages/core-graphql/lib/repositories/transactions.js]: `console.logs` don't belong into production code.

File: packages/core-graphql/lib/resolvers/queries/transaction/transactions.js [hunk 6]
```diff
@@ -8,11 +8,10 @@ const { transactions: repository } = require('../../../repositories')
  * @return {Transaction[]}
  */
 module.exports = async (root, args) => {
-  const { orderBy, filter } = args
-
+  const { orderBy, filter, limit } = args
   const order = formatOrderBy(orderBy, 'timestamp:desc')
-
-  const result = await repository.findAll({ ...filter, orderBy: order })
+console.log(limit)
```
Comment (faustbrian) [inline packages/core-graphql/lib/resolvers/queries/transaction/transactions.js]: `console.logs` don't belong into production code.

File: packages/core-graphql/__tests__/api/wallets.test.js [hunk 7]
```diff
@@ -66,10 +66,39 @@ describe('GraphQL API { wallets }', () => {
...
+      /**
```
Comment (faustbrian) [inline packages/core-graphql/__tests__/api/wallets.test.js]: This comment should be removed.

File: packages/core-graphql/__tests__/api/wallets.test.js [hunk 8]
```diff
@@ -66,10 +66,39 @@ describe('GraphQL API { wallets }', () => {
...
 
   describe.skip('GraphQL queries for Wallets - testing relationships', () => {
+    it('should verify that relationships are valid', async () => {
+      const query = '{ wallets(limit: 1) { transactions { id } } }'
+      const response = await utils.request(query)
+
+      expect(response).toBeSuccessfulResponse()
 
+      console.log(response.data)
+      /**
```
Comment (faustbrian) [inline packages/core-graphql/__tests__/api/wallets.test.js]: This comment should be removed.

**Comment by JeremiGendron** [comment]
I found it a bit hard to implement the Address type and also saw all but
one type use String; would it be OK to verify in the resolver instead of
the typedef? @faustbrian

On Nov 1, 2018 11:37 PM, "faustbrian" <notifications@github.com> wrote:

*@faustbrian* requested changes on this pull request.
------------------------------

In packages/core-graphql/__tests__/api/wallets.test.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262000>:

>
+      console.log(response.data)

Should be either removed or be replaced by an assertion.
------------------------------

In packages/core-graphql/__tests__/api/wallets.test.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262014>:

>
+      console.log(response.data)
+      /**
+       *  { data: { wallets: [ [Object] ] },
+            errors: [ { message: 'database.createCondition is not a function',
+              locations: [Array],
+              path: [Array],
+              extensions: [Object] } ]
+          }
+       */
+
+      console.log(response.data.wallets)

Should be either removed or be replaced by an assertion.
------------------------------

In packages/core-graphql/lib/defs/inputs.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262115>:

> @@ -12,7 +12,6 @@ module.exports = `
   scalar JSON
   scalar Limit
   scalar Offset
-  scalar Address

Address is the correct type. You can find it in
core-graphql/lib/resolvers/index.js.
------------------------------

In packages/core-graphql/lib/defs/inputs.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262121>:

> @@ -35,7 +34,7 @@ module.exports = `
     fee: Float
     blockId: String
     senderPublicKey: String
-    recipientId: Address
+    recipientId: String

Address is the correct type. You can find it in
core-graphql/lib/resolvers/index.js.
------------------------------

In packages/core-graphql/lib/repositories/transactions.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262156>:

> @@ -47,6 +51,7 @@ class TransactionsRepository extends Repository {

     applyConditions([selectQuery, countQuery])

+console.log(parameters.limit)

console.logs don't belong into production code.
------------------------------

In packages/core-graphql/lib/resolvers/queries/transaction/transactions.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262161>:

>    const order = formatOrderBy(orderBy, 'timestamp:desc')
-
-  const result = await repository.findAll({ ...filter, orderBy: order })
+console.log(limit)

console.logs don't belong into production code.
------------------------------

In packages/core-graphql/__tests__/api/wallets.test.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262183>:

>
+      console.log(response.data)
+      /**
+       *  { data: { wallets: [ [Object] ] },
+            errors: [ { message: 'database.createCondition is not a function',
+              locations: [Array],
+              path: [Array],
+              extensions: [Object] } ]
+          }
+       */
+
+      console.log(response.data.wallets)
+      /**

This comment should be removed.
------------------------------

In packages/core-graphql/__tests__/api/wallets.test.js
<https://github.com/ArkEcosystem/core/pull/1271#discussion_r230262192>:

>
+      console.log(response.data)
+      /**

This comment should be removed.

—
You are receiving this because you authored the thread.

Reply to this email directly, view it on GitHub
<https://github.com/ArkEcosystem/core/pull/1271#pullrequestreview-170960754>,
or mute the thread

**Comment by faustbrian** [comment]
1. Why is it hard to implement the `Address` type since it is already implemented?
2. What address fields are using `String` as type instead of `Address`?
3. Why would you verify in the resolver instead of the typedef?

**Comment by JeremiGendron** [comment]
1. It wasn't intuitive how a query should be formatted and the
implementation was incomplete (regex isn't a valid field of
GraphQLTypes.String afaict)
2. In the typedef, Wallet has address: String for example. I think it's the
only place an address is 'used' like that (the implementation doesn't work
out of the box and it's non-obvious why there's an Address type but no
publicKey or vendorfield types).
3. I found it easy to implement: have a regex.test with the address string
and throw an error with a clear message if it's not good (bubbles up nicely
at least in Jest).

I also asked for help initially for this reason (address scalar, how to
present in query), turns out there were more issues (actual bugs) that
needed to be fixed. I admit the Address scalar type should be implemented I
simply really can't find the proper syntax based off documentation and
error messages.

**Comment by JeremiGendron** [comment]
Cleaned it up a lot, also got Address to work (the regex used /g and wasn't working for that reason)

**Pull request merged**